cpp/src/io/parquet/writer_impl.cu

```diff
@@ -886,8 +890,80 @@ void writer::impl::init_encoder_pages(hostdevice_2dvector<gpu::EncColumnChunk>&
   stream.synchronize();
 }
 
+void snappy_compress(device_span<gpu_inflate_input_s> comp_in,
+                     device_span<gpu_inflate_status_s> comp_stat,
+                     size_t max_page_uncomp_data_size,
+                     rmm::cuda_stream_view stream)
+{
+  size_t num_comp_pages = comp_in.size();
+  do {
```

Comment:
I still think it would be better to use exceptions here, as such:

- replace the while loop with a try block
- replace the breaks with throws
- place the code after the while loop in the catch block
- remove the return from the end of the loop

AFAICT, it implements the same logic in a much cleaner way